components/model_analyzer/TorchBenchAnalyzer.py

```diff
@@ -0,0 +1,120 @@
+from components.model_analyzer.tb_dcgm_types.gpu_power_usage import GPUPowerUsage
+from components.model_analyzer.tb_dcgm_types.gpu_free_memory import GPUFreeMemory
+from components.model_analyzer.tb_dcgm_types.gpu_used_memory import GPUUsedMemory
+from components.model_analyzer.tb_dcgm_types.gpu_fp32active import GPUFP32Active
+from components.model_analyzer.tb_dcgm_types.record_aggregator import RecordAggregator
+from components.model_analyzer.tb_dcgm_types.tb_logger import set_logger, LOGGER_NAME
+from components.model_analyzer.tb_dcgm_types.config import *
+from components.model_analyzer.tb_dcgm_types.config import DEFAULT_MONITORING_INTERVAL
+
+import logging
+logger = logging.getLogger(LOGGER_NAME)
+
+class ModelAnalyzer:
+    def __init__(self):
+        self.gpu_factory = GPUDeviceFactory()
+        self.gpus = self.gpu_factory.verify_requested_gpus(['all', ])
+        # the metrics to be collected
+        # self.gpu_metrics = [GPUUtilization, GPUPowerUsage,
+        #                     GPUFreeMemory, GPUUsedMemory, GPUFP32Active, GPUTensorActive]
+        self.gpu_metrics = [GPUFP32Active]
+        # the final metric results. Its format is {GPU_UUID: {GPUUtilization: }}
+        self.gpu_metric_value = {}
+        self.gpu_monitor = None
+        self.gpu_records = None
+        self.config = AnalayzerConfig()
+        set_logger()
+
+    def start_monitor(self):
+        try:
+            self.gpu_monitor = DCGMMonitor(
+                self.gpus, self.config.monitoring_interval, self.gpu_metrics)
+            self.gpu_monitor.start_recording_metrics()
+        except TorchBenchAnalyzerException:
+            self._destory_monitor()
+            raise
+
+    def _destory_monitor(self):
+        self.gpu_monitor.destroy()
+        self.gpu_monitor = None
+    
+    def stop_monitor(self):
+        """
+        @return: collected gpu records @todo: add return type 
+        """
+        self.gpu_records = self.gpu_monitor.stop_recording_metrics()
+        self._destory_monitor()
+        return self.gpu_records
+    
+    def aggregate(self):
+        gpu_record_aggregator = RecordAggregator()
+        gpu_record_aggregator.insert_all(self.gpu_records)
+        # @Yueming Hao todo: groupby GPU, then metric. Then calculate the flops
+        records_groupby_gpu = {}
+        records_groupby_gpu = gpu_record_aggregator.groupby(
+            self.gpu_metrics, lambda record: record.device_uuid())
+        
+        for gpu in self.gpus:
+            self.gpu_metric_value[gpu.device_uuid()] = {}
+        for metric_type, metric in records_groupby_gpu.items():
+            for gpu_uuid, metric_value in metric.items():
+                self.gpu_metric_value[gpu_uuid][metric_type] = metric_value
+    
+    def set_monitoring_interval(self, attempted_interval):
+        """
+        The default monitoring internval is 10ms.
+        """
+        if attempted_interval < 0.001:
```

Comment:
i thought 100 ms is finest sampling rate? 

from https://docs.nvidia.com/datacenter/dcgm/latest/dcgm-user-guide/feature-overview.html

By default, DCGM provides the metrics at a sample rate of 1Hz (every 1000ms). Users can query the metrics at any configurable frequency (minimum is 100ms) from DCGM